### PR TITLE
Better altitude calculations

### PR DESCRIPTION
**Pull request by th3w4y**
https://github.com/PokemonGoF/PokemonGo-Bot/pull/5184

### CODE REVIEW

**Comment by th3w4y** [comment]
If possible when you rebase maybe do not override my commits :) so peoples can blame me for my work

File: pokemongo_bot/walkers/step_walker.py [hunk 1]
```diff
@@ -82,6 +82,6 @@ def get_next_position(self, origin_lat, origin_lng, origin_alt, dest_lat, dest_l
 
         next_lat = direct["lat2"]
         next_lng = direct["lon2"]
-        next_alt = origin_alt + (travel / total_distance) * (dest_alt - origin_alt) + random_alt_delta()
+        next_alt = dest_alt + random_alt_delta()
```
Comment (Anakin5) [inline pokemongo_bot/walkers/step_walker.py]: We should not remove this one, the original formula seems correct to me.

**Comment by Anakin5** [comment]
The install of scipy fail for me, do we have to run something special ?

**Comment by Anakin5** [comment]
Looks like scipy cannot really be install from pip on windows 7 64 bits. I cannot run the code :s

**Comment by th3w4y** [comment]
@Anakin5  i see let me see... not even an older version?

**Comment by th3w4y** [comment]
do you have the error log?

**Comment by Anakin5** [comment]
Cannot build some dependencies
`
      caller_level = 2)
    File "c:\python27\lib\site-packages\numpy\distutils\misc_util.py", line 972, in get_subpackage
      caller_level = caller_level + 1)
    File "c:\python27\lib\site-packages\numpy\distutils\misc_util.py", line 909, in _get_configuration_from_setup_py
      config = setup_module.configuration(*args)
    File "scipy\linalg\setup.py", line 20, in configuration
      raise NotFoundError('no lapack/blas resources found')
  numpy.distutils.system_info.NotFoundError: no lapack/blas resources found

  ----------------------------------------
  Failed building wheel for scipy
`

**Comment by th3w4y** [comment]
https://www.microsoft.com/en-us/download/details.aspx?id=44266

**Comment by th3w4y** [comment]
but i guess it is a problem even travis does not have the library... ok i guess i will research if there are wheels for linux mac os x and windows of some older version....

**Comment by th3w4y** [comment]
can you try pip install scipy==0.16.1

**Comment by Anakin5** [comment]
Same problem with 0.16.1
What about that ?
```
from geopy.distance import great_circle

def get_alt(self, at_point=None):
        if at_point is None:
            at_point = self._last_pos
        
        return sorted([(great_circle(at_point, k).meters, v) for k, v in self._elevation_at_point.items()])[0][1]
```
;)

File: pokemongo_bot/walkers/step_walker.py (hunk 1 above)
Comment (th3w4y) [inline pokemongo_bot/walkers/step_walker.py]: yes but would be handled in .get_alt() now!

**Comment by th3w4y** [comment]
to get the last 2 points?

**Comment by th3w4y** [comment]
10 times slower but who cares!

```

In [12]: %timeit get_alt(p)
1000 loops, best of 3: 1.68 ms per loop

In [13]: 
```

**Comment by Anakin5** [comment]
It just get the nearest. Yes it is slower but it is not a significant time. We only call that function once per second and it is also very simple.
I am thinking we cannot really ask everybody to do manual installation of scipy right ?

**Comment by th3w4y** [comment]
I will adopt my change to it... I am annoyed that the scipy does not worked on travis and on windows

**Comment by th3w4y** [comment]
I will give-up on scipy it would be a nightmare to explain to users how to install it...!

**Comment by th3w4y** [comment]
I can live with 33ms!


```
In [4]: %timeit test_polyline()
10 loops, best of 3: 33.3 ms per loop

In [5]: def test_polyline():
   ...:     for x in p._points:
   ...:         q = p.get_alt(at_point=x)
   ...:         

In [6]: 

In [6]: %timeit test_polyline()

10 loops, best of 3: 33.2 ms per loop

In [7]: 

In [7]: 
```

**Comment by Anakin5** [comment]
There is aproblem as well, the bot position is not initiliazed in the test cases:
`self.bot.position = [ex_orig[0], ex_orig[1], self.polyline.get_alt(ex_orig)]`
that is why the elevation at starting point is considered zero and calculation is wrong.

File: pokemongo_bot/walkers/step_walker.py (hunk 1 above)
Comment (th3w4y) [inline pokemongo_bot/walkers/step_walker.py]: and for the random delta when there is no altitude that random_delta is good enough

**Comment by th3w4y** [comment]
build is passing https://travis-ci.org/th3w4y/PokemonGo-Bot/builds/158224357

**Comment by th3w4y** [comment]
@Anakin5 it is initialised 

self.bot.position = [47.1706378, 8.5167405, 0]

0 is because you started with some crap from outside the polyline!
then if you reached destination... you should be at ex_orig elevation correct?

>There is aproblem as well, the bot position is not initiliazed in the test cases:
self.bot.position = [ex_orig[0], ex_orig[1], self.polyline.get_alt(ex_orig)]
that is why the elevation at starting point is considered zero and calculation is wrong.

**Comment by th3w4y** [comment]
but it is all fixed in .get_alt()
and since that is the case... you don't need geometry in 

```
-        next_alt = origin_alt + (travel / total_distance) * (dest_alt - origin_alt) + random_alt_delta()
+        next_alt = dest_alt + random_alt_delta()
```
with the PolylineWalker you get polyline .get_alt() + random_alt_delta()

with the StepWalker you get uniform(self.bot.config.alt_min, self.bot.config.alt_max) + random_alt_delta()

**Comment by Anakin5** [comment]
I have a complete different view on the problem. We need to take everything back to zero I think because we are not making the same assumptions.

**Comment by th3w4y** [comment]
I believe once you have a polyline!... you should immediately adopt the elevation provided by it not drift from 0 to 432... 

phone can return 0 by euro but does not drift from 0 to 432 in baby steps...

**Comment by th3w4y** [comment]
In Your approach the first polyline you'll walk you will drift from 0 (or whatever uniform) to destination elevation in baby steps during the walk of first polyline until pokestop!

**Comment by Anakin5** [comment]
I commit my changes so that you can see what I am thinking

**Comment by th3w4y** [comment]
@Anakin5 i already know!

**Comment by th3w4y** [comment]
You are using:

`origin_alt + (travel / total_distance) * (dest_alt - origin_alt) + random_alt_delta()`

to slowly reach the alt destination in baby steps!

**Comment by Anakin5** [comment]
yes, what is wrong with that ?

**Comment by th3w4y** [comment]
if there no directions! you fallback to StepWalker behaviour!!! you are presented with 2 point and no altitude... you will jump back to uniform(min_alt, max_alt) in one go!

**Comment by th3w4y** [comment]
my argument is this:

phone can return 0 and small value for elevation by error but does not drift from 0 to 432 in baby steps...

**Comment by Anakin5** [comment]
yes but I still don't get it. where does this zero comes from ?

**Comment by th3w4y** [comment]
how is the bot initialised?

**Comment by Anakin5** [comment]
Can you just explain why you don't want the elevation to be proportional to the distance walk between the origin and the destination ?
Because I stuck on that question so I cannot follow the rest of the reasoning :)

**Comment by th3w4y** [comment]
Reasons:

- initial altitude is 0 because of geoy not supporting elevations
- phone does not drift from 0 to 430 in 10minutes walk (imagine first polyline around 1km)
- with the trigonometry formula you get +/- random_alt_delta() at each step() [I don't understand why you don't want that?]
- i want to have good elevation as much as possible from start of bot till end of bot... not wait to walk first polyline to synchronise

**Comment by Anakin5** [comment]
The problem you are mentioning has little to do with the formula. It is about initializing the elevation to something not zero. The formula is still correct between 2 polyline points.

Now to address the problem of the first position, that is not here that the fix is. The position is sent at the very start of the bot, we are not even in the Walker task yet. So the problem will not get fixed here.
We could se the altitude to the first point when we see it is currently zero but jumping from 0 to something is not a very good solution as well.

Let's open a PR to fix the initial altitude, ok ?

**Comment by th3w4y** [comment]
what if your next pokestop is too close for google direction api... you get 2 point your altitude drops down again and only the next valid polyline you get to dript again from uniform(min, max) to destination of the polyline elevation in baby steps

**Comment by Anakin5** [comment]
We never use the random altitude in Polyline. Where do you see that the value returned by uniform is used ?

**Comment by th3w4y** [comment]
this will happen when you have big density of close to eachother pokestops

actually i don't want StepWalker to change the altitude at all it it was set by the polyline at one point, meaning Polyline walker should .get_alt() or bot.position[2]

**Comment by Anakin5** [comment]
yes that is what it is doing. It does not use the random value.

**Comment by th3w4y** [comment]
it does bevcause the pilyline does not give you an aktitude...

**Comment by Anakin5** [comment]
`origin_lat, origin_lng, origin_alt = self.bot.position`
`dest_alt = polyline.get_alt() or origin_alt`

Please copy me the lines that show that it does use the random value

**Comment by th3w4y** [comment]
if dest_alt is None:
            self.dest_alt = uniform(self.bot.config.alt_min, self.bot.config.alt_max)
        else:
            self.dest_alt = dest_alt

**Comment by th3w4y** [comment]
you are assumin that people configured that correctly!

**Comment by th3w4y** [comment]
in StepWalker is that code....

**Comment by Anakin5** [comment]
configure what ? the initial value of dest_alt is never used. Agin, please follow what we do with self.dest_all and you will see.

**Comment by Anakin5** [comment]
```
        if dest_alt is None:
            self.dest_alt = uniform(self.bot.config.alt_min, self.bot.config.alt_max)
        else:
            self.dest_alt = dest_alt
```

```
        origin_lat, origin_lng, origin_alt = self.bot.position

        new_position = self.get_next_position(origin_lat, origin_lng, origin_alt, self.dest_lat, self.dest_lng, self.dest_alt, speed)
```

dest_alt overriden just here:
```
dest_alt = polyline.get_alt() or origin_alt
```

**Comment by Anakin5** [comment]
Just simply never used. Whatever is the initialisation value, it is just never used on the PolylineWalker.

**Comment by th3w4y** [comment]
ok so lets start... 
- bot starts elevation 0 (to be fixed)

now yoycreate a PolylineWalker object...
-  __init__ method from step walker is used self.dest_alt = uniform
- you call first step
- you reach the method in Polylinewalker is the polyline.get_alt() or orig_alt
- polyline has only 2 point get_alt() return None

What is your altitude?

**Comment by Anakin5** [comment]
orig_alt

**Comment by th3w4y** [comment]
is orig_alt = uniform(min, max)

**Comment by Anakin5** [comment]
no

**Comment by th3w4y** [comment]
from the __init__

**Comment by Anakin5** [comment]
it is self.bot.position[2]

**Comment by th3w4y** [comment]
from where?

**Comment by th3w4y** [comment]
show me line number maybe i am missing something...

**Comment by Anakin5** [comment]
step_walker line 37: `origin_lat, origin_lng, origin_alt = self.bot.position`

**Comment by Anakin5** [comment]
that is dest_alt which is initilized with unifrm. Not orig_alt. orig_alt is initilized with current bot position.

**Comment by th3w4y** [comment]
my bad... i missed that one mia culpa...

**Comment by Anakin5** [comment]
hehe :)

**Comment by Anakin5** [comment]
Now you can read all the code again with that in mind so you might end up thinking same as me, haha

**Comment by th3w4y** [comment]
so uou will be drifting down towards dest_alt

**Comment by th3w4y** [comment]
end of that polyline you are at uniform(min, max)

:))

**Comment by Anakin5** [comment]
... no

**Comment by th3w4y** [comment]
just kidding you have passed arround orig_alt

**Comment by th3w4y** [comment]
is ok

**Comment by Anakin5** [comment]
ok

**Comment by Anakin5** [comment]
altitude is proportional between current bot position altitude and destination altitude (which is either the one return by polyline or current). Are we ok with that ?

**Comment by th3w4y** [comment]
Then from all my arguments ( assuming we fix initializing elevation) there is just one remaining...

- None

**Comment by Anakin5** [comment]
Yes all is passing now, build is green

**Comment by th3w4y** [comment]
that oart i also don' really like cause is like going towards exponentially not folowing big natural elevations....

**Comment by th3w4y** [comment]
if i have. a 20m delat hill that i have to want over in like 3 minutes... the elevation bbased on proportion of total are really strange

**Comment by Anakin5** [comment]
I am lost again

**Comment by Anakin5** [comment]
what is 20m ? what is 3 minutes ?

**Comment by th3w4y** [comment]
I don't understand why you like your code more ( beside that is yours) since they both do the same.

but my method ensures you are always of by only +- randolm_alt_delta

**Comment by th3w4y** [comment]
I live is Switzerland we have hilll and mountains here, there are +/-50m alt up and down jumps in 200m

**Comment by Anakin5** [comment]
You method set the altitude as beeing the destination altitude. If current altitude is 10 and destination is 20 and you start walking from the first point, you imediately set it to 20, which put you off by 10 meters.

**Comment by th3w4y** [comment]
here in 430m 100m away is 470 100m further is again 410

**Comment by Anakin5** [comment]
The Polyline points are closer than that.

**Comment by Anakin5** [comment]
yes it is

**Comment by th3w4y** [comment]
is the elevations at point disctionary...

**Comment by th3w4y** [comment]
you get a value for each step()

**Comment by th3w4y** [comment]
i tale the final total_distance / speed and that many ebtries are in the elevation_at_point dict is more then point for direction

**Comment by Anakin5** [comment]
again I don't see what you want to explain to me. What altitude do you want to take if not from the elevation points that are in the Polyline class ?

**Comment by th3w4y** [comment]
check the init of Polyline

**Comment by th3w4y** [comment]
regarding the nr of samoles i fetch...

**Comment by Anakin5** [comment]
What altitude do you want to take if not from the elevation points that are in the Polyline class ?

**Comment by Anakin5** [comment]
You want to call get_alt in get_next_destination I think, is that what you mean ?

**Comment by th3w4y** [comment]
get_alt(next_point)

**Comment by Anakin5** [comment]
look what I just commit, do you prefer like that ?

**Comment by th3w4y** [comment]
polylinewalker should

return next_lat, next_lng, polyline.get_alt(next_lat, next_lng)

**Comment by Anakin5** [comment]
I made a second one

**Comment by Anakin5** [comment]
third one the good one !

**Comment by th3w4y** [comment]
@Anakin5 this is how it was in the morning

**Comment by Anakin5** [comment]
Test cases will have to to be modified to know what altitude we are expecting.

**Comment by th3w4y** [comment]
https://github.com/Anakin5/PokemonGo-Bot/pull/7

**Comment by th3w4y** [comment]
@Anakin5 please let me keep _get_alt and _get_relative_hight

**Comment by th3w4y** [comment]
That way i could use it in tests if needed be to identify cornet cases!

**Comment by Anakin5** [comment]
ok

**Comment by th3w4y** [comment]
I only fixed the no self.elevations_at_point and readied those as private methods...

**Pull request closed without merging**